gateway/src/metrics.rs

```diff
@@ -0,0 +1,1177 @@
+        //
+        // A broadcast channel is used here, not because we are actually
+        // multi-consumer (`Producer::produce` is never called concurrently),
+        // but because the broadcast channel has properly ring-buffer-like
+        // behavior, where earlier messages are discarded, rather than exerting
+        // backpressure on senders (as Tokio's MPSC channel does). This
+        // is what we want, as we would prefer a full buffer to result in
+        // clobbering the oldest measurements, rather than leaving the newest
+        // ones on the floor.
+        let (sample_tx, sample_rx) =
+            broadcast::channel(cfg.max_buffered_sample_chunks);
+
+        // Using a channel for this is, admittedly, a bit of an end-run around
+        // the `OnceLock` on the `ServerContext` that *also* stores the rack ID,
+        // but it has the nice benefit of allowing the `PollerManager` task to _await_
+        // the rack ID being set...we might want to change other code to use a
+        // similar approach in the future.
+        let (rack_id_tx, rack_id_rx) = oneshot::channel();
+        let rack_id_tx = if let Some(rack_id) = rack_id {
+            rack_id_tx.send(rack_id).expect(
+                "we just created the channel; it therefore will not be \
+                     closed",
+            );
+            None
+        } else {
+            Some(rack_id_tx)
+        };
+        let pollers = {
+            let log = log.new(slog::o!("component" => "sensor-poller"));
+            slog::info!(
+                &log,
+                "SP sensor metrics configured";
+                "poll_interval" => ?cfg.sp_poll_interval,
+                "max_buffered_sample_chunks" => cfg.max_buffered_sample_chunks,
+            );
+
+            tokio::spawn(
+                PollerManager {
+                    sample_tx,
+                    apictx,
+                    poll_interval: cfg.sp_poll_interval,
+                    tasks: tokio::task::JoinSet::new(),
+                    log,
+                    mgs_id: id,
+                }
+                .run(rack_id_rx),
+            )
+        };
+
+        let (addrs_tx, addrs_rx) =
+            tokio::sync::watch::channel(addresses.clone());
+        let server = {
+            let log = log.new(slog::o!("component" => "producer-server"));
+            let registry = ProducerRegistry::with_id(id);
+            registry
+                .register_producer(Producer { sample_rx, log: log.clone() })
+                .context("failed to register metrics producer")?;
+
+            tokio::spawn(
+                ServerManager { log, addrs: addrs_rx, registry }.run(cfg),
+            )
+        };
+        Ok(Self { addrs_tx, rack_id_tx, server, pollers })
+    }
+
+    pub fn set_rack_id(&mut self, rack_id: Uuid) {
+        if let Some(tx) = self.rack_id_tx.take() {
+            tx.send(rack_id).expect("why has the sensor-poller task gone away?")
```

Comment:
Nit: could this happen at Tokio shutdown, when tasks are destroyed in arbitrary order?